File: bot-core/src/main/kotlin/fr/delphes/bot/overlay/OverlayElementGeneralProperties.kt
package fr.delphes.bot.overlay

import kotlinx.serialization.Serializable

@Serializable
data class OverlayElementGeneralProperties(
    val id: String,
    val left: Int,
    val top: Int,
    val sortOrder: Int = 0,
) : Comparable<OverlayElementGeneralProperties> {
    override fun compareTo(other: OverlayElementGeneralProperties) = comparator.compare(this, other)

    companion object {
        private val comparator = Comparator.comparing(OverlayElementGeneralProperties::sortOrder)
    }
}

File: ui-overlay/src/overlay/OverlayElementGeneralProperties.ts
export class OverlayElementGeneralProperties {
  id: string
  left: number
  top: number
  sortOrder: number

  constructor(id: string, left: number, top: number, sortOrder: number) {
    this.id = id
    this.left = left
    this.top = top
    this.sortOrder = sortOrder
  }

  public static fromJson(json: {
    id: string
    left: number
    top: number
    sortOrder: number
  }): OverlayElementGeneralProperties {
    return new OverlayElementGeneralProperties(json.id, json.left, json.top, json.sortOrder)
  }
}